trackintel/preprocessing/trips.py

```diff
@@ -0,0 +1,433 @@
+        geom_col = None  # not used
+        crs_is_projected = False  # not used
+    else:
+        # if no location is given, we need the trips table to have a geometry column
+        assert isinstance(trips_inp, gpd.geodataframe.GeoDataFrame), "Trips table must be a GeoDataFrame"
+        geom_col = trips_inp.geometry.name
+        # get crs
+        crs_is_projected = ti.geogr.distances.check_gdf_crs(trips_inp.copy())
+
+    trips = trips_inp.copy()
+    # If the trips already have a column "tour_id", we drop it
+    if "tour_id" in trips:
+        trips.drop(columns="tour_id", inplace=True)
+        warnings.warn("Deleted existing column 'tour_id' from trips.")
+
+    # convert hours to timedelta
+    max_time_delta = timedelta(hours=max_time)
+
+    kwargs = {
+        "max_dist": max_dist,
+        "max_nr_gaps": max_nr_gaps,
+        "max_time": max_time_delta,
+        "staypoints": staypoints,
+        "geom_col": geom_col,
+        "crs_is_projected": crs_is_projected,
+    }
+    if print_progress:
+        tqdm.pandas(desc="User trip generation")
+        tours = (
+            trips.groupby(["user_id"], group_keys=False, as_index=False)
+            .progress_apply(_generate_tours_user, **kwargs)
+            .reset_index(drop=True)
+        )
+    else:
+        tours = (
+            trips.groupby(["user_id"], group_keys=False, as_index=False)
+            .apply(_generate_tours_user, **kwargs)
+            .reset_index(drop=True)
+        )
+
+    # No tours found
+    if len(tours) == 0:
+        return trips, tours
+
+    # index management
+    tours["id"] = np.arange(len(tours))
+    tours.set_index("id", inplace=True)
+
+    # assign tour id to trips
+    tour2trip_map = tours[["trips"]].to_dict()["trips"]
+    ls = []
+    # Each trip is only assigned to one tour. If a trip belongs to multiple tours, we can find its smallest subtour
+    # by using the first one it is assigned to (nested tours are always found before big tours - have smaller tour_id)
+    already_assigned = []
+    for key, values in tour2trip_map.items():
+        for value in values:
+            if value not in already_assigned:
+                ls.append([value, key])
+                already_assigned.append(value)
+    temp = pd.DataFrame(ls, columns=[trips.index.name, "tour_id"]).set_index(trips.index.name)
+    # temp_grouped = temp.groupby("id").agg({"tour_id": list}) # other version: provide list of tour ids
+    trips = trips.join(temp, how="left")
+
+    ## dtype consistency
+    # trips id (generated by this function) should be int64
+    tours.index = tours.index.astype("int64")
+    trips["tour_id"] = trips["tour_id"].astype("Int64")
+
+    # cleaning
+    # tours.drop(columns=["trips"], inplace=True)
+
+    return trips, tours
+
+
+def _generate_tours_user(
+    user_trip_df,
+    staypoints=None,
+    max_dist=100,
+    max_nr_gaps=0,
+    max_time=timedelta(days=1),
+    geom_col="geom",
+    crs_is_projected=False,
+):
```

Comment:
I can grasp the basic idea, but my head won't work recognizing the corner cases - -. I'll believe you and Henry until I started using this :).

What could also help is drawing a flowchart for the algorithm, like our old generate_trip function :). 

https://github.com/mie-lab/trackintel/blob/master/docs/_static/tripalgorithm.png

Reply:
We can do that for the trackintel paper ;) I'll put it on my todo list